upgrades/backend.go

```diff
@@ -42,3 +56,175 @@ func (s stateBackend) AddVirtualHostKeys() error {
 func (s stateBackend) SplitMigrationStatusMessages() error {
 	return state.SplitMigrationStatusMessages(s.pool)
 }
+
+// PopulateApplicationStorageUniqueID runs an upgrade to backfill CAAS apps
+// storage unique IDs.
+func (s stateBackend) PopulateApplicationStorageUniqueID() error {
+	return state.PopulateApplicationStorageUniqueID(s.pool, GetStorageUniqueIDs(NewK8sClient))
+}
+
+// NewK8sClient initializes a new k8s client for a given model.
+func NewK8sClient(model *state.Model) (kubernetes.Interface, *rest.Config, error) {
+	g := stateenvirons.EnvironConfigGetter{Model: model}
+	cloudSpec, err := g.CloudSpec()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	cfg, err := jujukubernetes.CloudSpecToK8sRestConfig(cloudSpec)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	k8sClient, err := kubernetes.NewForConfig(cfg)
+	return k8sClient, cfg, err
+}
+
+type newK8sFunc func(model *state.Model) (kubernetes.Interface, *rest.Config, error)
+
+type kubernetesClient struct {
+	kubernetes.Interface
+	config *rest.Config
+}
+
+// GetStorageUniqueIDs attempts to grab the storage unique ID for each app
+// in the given model.
+// The storage unique ID is saved in annotations in a statefulset and for legacy
+// applications are saved in a deployment or daemonset.
+func GetStorageUniqueIDs(newK8sClient newK8sFunc) func(
```

Comment:
Let's pass in a context here and have the caller use `context.Background)` if needed